### PR TITLE
Split piwik.js in multiple files minor tweak

### PR DESCRIPTION
**Pull request by mattab**
refs #10441

### CODE REVIEW

File: plugins/CustomPiwikJs/File.php [hunk 1]
```diff
@@ -60,7 +60,7 @@ public function getName()
      */
     public function hasWriteAccess()
     {
-        return $this->hasReadAccess() && is_writable($this->file);
+        return is_writable($this->file);
```
Comment (mattab) [inline plugins/CustomPiwikJs/File.php]: hopefully it's fine to just check whether file is writable. It solves the issue when the target file does not exist yet, hasWriteAccess should return true when  the file can be created/written to.

Comment (tsteur) [inline plugins/CustomPiwikJs/File.php]: FYI: I had it like this on purpose for now and in tests just created the file in advance

**Comment by tsteur** [comment]
We now need to update the tests that create this file in advance and also `tests/javascript/index.php` which creates this file in advance too `file_put_contents($targetFile, '');`

**Comment by mattab** [comment]
After removing the file_put_contents the JavaScript tests fail with 
```
Executing tests in test suite JavascriptTests...

Phantom callbacks not attached in time.  See http://github.com/mark-rushakoff/OpenPhantomScripts/issues/1

Unsafe JavaScript attempt to access frame with URL about:blank from frame with URL file://../javascript/testrunner.js. Domains, protocols and ports must match.

```

I'm wondering whether it's caused by my change or not. I'll try to re-add the file_put_contents again and see

**Comment by mattab** [comment]
Works if I leave the file_put_contents so I'll leave it for now...

**Comment by tsteur** [comment]
If it works when you leave the file_put_contents, this PR (https://github.com/piwik/piwik/pull/10449/files#diff-296a183c286a6fe5a8c5d1ae4ecea1d3L63) seems to not 100% work

**Comment by tsteur** [comment]
Be good to add a test for this as I don't think it works. I think there might be even a test that would have failed otherwise if it worked so no test would have needed but not sure

**Comment by mattab** [comment]
@tsteur I'm a bit confused here... can't think of a case where it does not work. willing to leave it as is unless you point me to the issue (hope i'm not lazy)

**Comment by tsteur** [comment]
If your PR worked it would fail eg https://github.com/piwik/piwik/blob/2.x-dev/plugins/CustomPiwikJs/tests/Integration/FileTest.php#L69 

See http://us2.php.net/manual/en/function.is-writable.php `is_writable` only returns true when file exists and is writable.  What you want is probably `is_writable(dirname($this->file))` but recursive up in case the directory is missing but one of the other directories is missing and it needs to create the directories when calling `save()`. Eg it might be `/home/test/foo/bar/baz` in this case it needs to first test `/home/test/foo/bar/`, then `/home/test/foo/` then `/home/test/foo/` and on `->save()` create missing directories

**Comment by mattab** [comment]
> If your PR worked it would fail eg https://github.com/piwik/piwik/blob/2.x-dev/plugins/CustomPiwikJs/tests/Integration/FileTest.php#L69

This test should only fail when the `NOT_EXISTING_FILE`'s directory is not writable. on travis and locally the directory is writable so even though file doesn't exist, it's still writable and can be created by the console command. that's actually the reason I removed the `is_readable`: so that when the directory is writable but the file doesn't exist yet, the console won't complain. Am I missing something?

**Comment by mattab** [comment]
Ok now it's clear, thanks, see the PR that fixes the issue https://github.com/piwik/piwik/pull/10576/files

**Pull request merged**